### PR TITLE
Support changing dimension schema in Auto Compaction 

### PR DESCRIPTION
**Pull request by maytasm**
Support changing dimension schema in Auto Compaction 

### Description

Changing dimensionsSpec is already supported in manual compaction task. This PR adds it to the auto compaction.
By changing dimensionsSpec, the user will be able to change the dimension ordering, remove dimension, and change dimension data type. 

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by suneet-s — APPROVED** [review]
+1 after CI

File: docs/configuration/index.md [hunk 1]
```diff
@@ -988,7 +989,7 @@ The below is a list of the supported configurations for auto compaction.
 |`chatHandlerNumRetries`|Retries for reporting the pushed segments in worker tasks.|no (default = 5)|
 
 ###### Automatic compaction granularitySpec
-You can optionally use the `granularitySpec` object to configure the segment granularity of the compacted segments.
+You can optionally use the `granularitySpec` object to configure the granularity and rollup of the compacted segments.
```
[the remark this answers is not in the document]
Comment (maytasm) [reply docs/configuration/index.md:992]: Done

File: core/src/test/java/org/apache/druid/timeline/DataSegmentTest.java [hunk 2]
```diff
@@ -120,6 +121,7 @@ public void testV1Serialization() throws Exception
         new NumberedShardSpec(3, 0),
         new CompactionState(
             new HashedPartitionsSpec(100000, null, ImmutableList.of("dim1")),
+            new DimensionsSpec(DimensionsSpec.getDefaultSchemas(ImmutableList.of("dim1", "bar", "foo")), null, null),
```
[the remark this answers is not in the document]
Comment (maytasm) [reply core/src/test/java/org/apache/druid/timeline/DataSegmentTest.java:124]: Done. See DataSegmentTest#testDeserializationDataSegmentLastCompactionStateWithoutDimensionsSpec

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 3]
```diff
@@ -440,6 +442,19 @@ private boolean needsCompaction(DataSourceCompactionConfig config, SegmentsToCom
...
 
+    if (config.getDimensionsSpec() != null) {
+      final DimensionsSpec existingDimensionsSpec = lastCompactionState.getDimensionsSpec();
+      // Checks for list of dimensions
+      if (config.getDimensionsSpec().getDimensions() != null) {
+        final List<DimensionSchema> existingDimensions = existingDimensionsSpec != null ?
+                                                         existingDimensionsSpec.getDimensions() :
+                                                         null;
+        if (!config.getDimensionsSpec().getDimensions().equals(existingDimensions)) {
+          return true;
```
[the remark this answers is not in the document]
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:453]: Done

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 4]
```diff
@@ -440,6 +442,19 @@ private boolean needsCompaction(DataSourceCompactionConfig config, SegmentsToCom
       }
     }
 
+    if (config.getDimensionsSpec() != null) {
+      final DimensionsSpec existingDimensionsSpec = lastCompactionState.getDimensionsSpec();
+      // Checks for list of dimensions
+      if (config.getDimensionsSpec().getDimensions() != null) {
```
[the remark this answers is not in the document]
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:448]: Yes. If `config.getDimensionsSpec().getDimensions() == null` then needsCompaction will return false regardless of lastCompactionState.getDimensionsSpec() value

**Pull request merged**